### PR TITLE
✨ support server group and scheduler hint additional properties

### PR DESCRIPTION
**Pull request by okozachenko1203**
✨ 
**What this PR does / why we need it**:
This PR adds a new field `SchedulerHintAdditionalProperties` in both `OpenstackServerSpec` and `OpenstackMachineSpec` to specify nova scheduler hint properties during server creation.
In addition, it passes `openStackMachineSpec.ServerGroup` to `OpenStackServerSpec.ServerGroup` which seems missing during the spec conversion.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**TODOs**:

- [ ] squashed commits
- if necessary:
  - [ ] includes documentation
  - [ ] adds unit tests

/hold

### CODE REVIEW

**Comment by jichenjc** [comment]
>>>This PR adds a new field SchedulerHintAdditionalProperties in both OpenstackServerSpec and OpenstackMachineSpec to specify nova scheduler hint properties during server creation.
In addition, it passes openStackMachineSpec.ServerGroup to OpenStackServerSpec.ServerGroup which seems missing during the spec conversion.

I think it make sense to split those 2 as 2nd seems to be a bug and #1 to be a feature

File: api/v1alpha1/openstackserver_types.go
Comment (jichenjc) [inline api/v1alpha1/openstackserver_types.go:110]: seems this is not true, we can't expect it is always a string from the API spec?

https://docs.openstack.org/api-ref/compute/

```


    The request validation schema is per hint. For example, some require a single string value, and some accept a list of values.

    Hints are only used based on the cloud scheduler configuration, which varies per deployment.

    Hints are pluggable per deployment, meaning that a cloud can have custom hints which may not be available in another cloud.

For these reasons, it is important to consult each cloud’s user documentation to know what is available for scheduler hints.
```

Comment (okozachenko1203) [reply api/v1alpha1/openstackserver_types.go:110]: @jichenjc yes, it is true.
strictly speaking, `SchedulerHintAdditionalProperties` presents `os:scheduler_hints["additionalProperties"]`, But it is also any type. My concern is i cannot use `map[string]interface{}` because kubebuilder doesn't support it.
I couldn't find any other reference or type struct def which handles `any` type, and i used a `string` here because our custom filter accepts boolean and string type values.

Comment (mnaser) [reply api/v1alpha1/openstackserver_types.go:110]: @jichenjc do you have thought on this? :)

Comment (jichenjc) [reply api/v1alpha1/openstackserver_types.go:110]: no solution but just questioning the API and the way of implementation
if we are sure the string (maybe define our own format) , or we claim we only support string type filter
it might be ok...

File: pkg/cloud/services/compute/instance.go
Comment (jichenjc) [inline pkg/cloud/services/compute/instance.go:122]: and as group is part of the scheduler properties ,should we consider consolidate them?

Comment (okozachenko1203) [reply pkg/cloud/services/compute/instance.go:122]: @jichenjc well, it is true that all those fields in `servers.schedulerHintOpts` are merged and passed as `os:scheduler_hints` request body parameter, so we can set `group` under `additionalProperties`. But I am not sure if that is necessary because gophercloud already defined `Group` as an explicit parameter (i think it is just for convenient) under `servers.SchedulerHintOpts`. 
https://github.com/gophercloud/gophercloud/blob/5cb81d730a1e027aa5981edd23c7b1403411db40/openstack/compute/v2/servers/requests.go#L141

File: api/v1alpha1/openstackserver_types.go
Comment (okozachenko1203) [reply api/v1alpha1/openstackserver_types.go:110]: i am not sure if any changes required for this convo

**Comment by mdbooth** [comment]
> /cherry-pick release-0.10 /cherry-pick release-0.9

We would not typically backport this kind of change. Backporting to 0.9 in particular would likely be very difficult.

**Review by mdbooth — COMMENTED** [review]
As noted below we should convert this map into a list in the API, and marshal it into an appropriate map before send it to gophercloud.

I'm also pontificating on @jichenjc's observation that the `value` type should be `any` rather than string. I also note that your own custom filter accepts boolean or string. Another might accept an arbitrary struct?

How about using a discriminated union here:
```yaml
additionalProperties:
- name: my_custom_bool_hint
  value:
    type: bool
    bool: true
- name: my_custom_string_hint
  value:
    type: string
    string: "foo"
- name: my_custom_int_hint
  value:
    type: int
    int: 7
- name: my_custom_object_hint
  value:
    type: object
    object:
      foo: bar
      baz:
      - baz1
      - baz2
```

I'd recommend against implementing any types that we don't currently need, but `object` would probably involve RawExtension in some capacity. This would be unmarshalled into `map[string]any` before passing to gophercloud.

@JoelSpeed got a better idea?

File: pkg/cloud/services/compute/instance.go
Comment (mdbooth) [reply pkg/cloud/services/compute/instance.go:122]: I think it's correct to leave Group separate here. I considered asking that we validate that AdditionalProperties does not contain `group`, but the problem with that is that we can't easily extend that in the future as we add new 'supported' hints.

So I think additionalProperties is a footgun, but there's no good way round that. We should just document that adding 'group' to it is a bad idea and has undefined behaviour.

**Review by mdbooth — COMMENTED** [review]
Wow, thanks for coming back on this so quickly.

The change from map to array is definitely good. Although it was my suggestion, I'm still hoping we can come up with something better than the union discriminated by type for the value. If I can find a minute I'm going to protype replacing the whole value with a RawExtension and write some tests to see what that looks like in practise. If it's ok that might be a better solution.

If we do go with the discriminated union, I'd say:
* lets omit `object` for now because I'm going to ask for validation and tests, and you don't want to write those for something nobody's using yet
* lets rename `int` to `number` so our janky JSON marshaller at least uses the correct JSON jargon

Whatever we come up with, we will need to write CEL to validate it, and test those validations in the apivalidations suite. They're pretty easy to write and there are a ton of examples to follow already.

I'll try to prototype the RawExtension thing today, but if I'm taking too long feel free to try it yourself. I want to know:
* What does it look like to populate in Go
* What does it look like to populate in a yaml template
* Are there any quirks in how it marshals to json which would make it unusable for this?

Ideally it just acts like `any`, but I don't know.

File: controllers/openstackmachine_controller.go
Comment (okozachenko1203) [reply controllers/openstackmachine_controller.go:483]: I think so

File: api/v1beta1/openstackmachine_types.go
Comment (mdbooth) [inline api/v1beta1/openstackmachine_types.go:41]: Please can you define this as an enum, both in Go and in the CRD. Example: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/90da86dc16a85a6100a07ab3a5a9d70389ddb8bb/api/v1beta1/types.go#L546-L554

File: api/v1beta1/openstackmachine_types.go
Comment (mdbooth) [inline api/v1beta1/openstackmachine_types.go:53]: s/Int/Number/

File: api/v1beta1/openstackmachine_types.go
Comment (mdbooth) [inline api/v1beta1/openstackmachine_types.go:44]:
```suggestion
	// This field is required if type is 'Bool', and must not be set otherwise.
```

And similar for other fields.

File: api/v1beta1/openstackmachine_types.go
Comment (mdbooth) [inline api/v1beta1/openstackmachine_types.go:40]: These are typically capitalised in kube APIs.

**Review by mdbooth — COMMENTED** [review]
This looks great, thanks. Incidentally, I tested out RawExtension and it's not what we want: it requires the data to have an APIVersion and Kind, which isn't relevant to this use case. Based on my quick tinkering I think it will be possible to create an `Object` type, but until we have somebody who needs it I don't think it's worth the effort to write. It will certainly be more complex than the simple types implemented here.

Please can you add API validations for the new API. I've linked examples below and it should be fairly straightforward. Please can you also add tests to `test/e2e/suites/apivalidations/openstackmachine_test.go`. There are lots of examples in there and these should be very straightforward. I think we want:
* empty list is ok
* item with empty name is not ok
* item with empty value is not ok
* value.type is `bool` and `value.bool` is populated is ok
* value.type is `bool` and `value.string` is populated is not ok
* ... and repeat

Please can we also add something to the unit tests, probably https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/90da86dc16a85a6100a07ab3a5a9d70389ddb8bb/pkg/cloud/services/compute/instance_test.go#L409-L410 to validate that we're marshalling the struct correctly when sending it to gophercloud? We're already passing expectedSchedulerHintOpts into expectCreateServer, so this should be fairly easy to add:
* Modify expectCreateServer to take servers.SchedulerHintOpts instead of SchedulerHintOptsBuilder, as this will be easier to validate. This change shouldn't break anything.
* Add a new test: "With custom scheduler hint number" (or whatever)
* In getInstanceSpec, fetch the default, set the required scheduler hints
* In expect, call expectCreateServer with the expected schedulerHintOpts.

**Review by jichenjc — COMMENTED** [review]
/lgtm

**Review by jichenjc — COMMENTED** [review]
/lgtm

File: test/e2e/suites/apivalidations/openstackmachine_test.go
Comment (mdbooth) [inline test/e2e/suites/apivalidations/openstackmachine_test.go:438]: nit: could you make these separate tests with their own It()? This both allows them to run in parallel, and also enables ginkgo to separate out the logs if one of them fails.

**Review by mdbooth — COMMENTED** [review]
This is fantastic work, thanks. I wouldn't block on any of the nits, but please can you:
* Add the 2 extra unit test (these are just cut/paste)
* Squash commits.

If nobody else has any new concerns I'll be happy to approve.

File: api/v1beta1/openstackmachine_types.go
Comment (EmilienM) [inline api/v1beta1/openstackmachine_types.go:85]: did we miss the validation?

```
// +kubebuilder:validation:Required
```

**Review by EmilienM — APPROVED** [review]
There are enough "nits" to request a follow-up, whether in this PR or after it has merged.
Thanks for this work!

/approve

File: api/v1beta1/openstackmachine_types.go
Comment (mdbooth) [reply api/v1beta1/openstackmachine_types.go:85]: It's implicit here, which is why the tests pass, but I agree that we should add it anyway. We've been trying to be explicit with these.

**Comment by EmilienM** [comment]
/lgtm
thanks for addressing the comments.

Feel free to hold cancel once CI jobs have passed.

**Comment by jichenjc** [comment]
um... looks like the CI for the pull-cluster-api-provider-openstack-test stuck somewhere..

**Comment by EmilienM** [comment]
Job is running. Since it's required we can hold cancel and it'll land once the CI job has finished.
Thanks for all the work here and your patience.

/hold cancel

**Comment by EmilienM** [comment]
ugh it went sleep again
/test pull-cluster-api-provider-openstack-test

**Comment by EmilienM** [comment]
@okozachenko1203 I don't know what's going on in CI but could you please try a rebase on `main`, this will trigger jobs. I'll add the LGTM again.

File: api/v1beta1/openstackmachine_types.go
Comment (JoelSpeed) [reply api/v1beta1/openstackmachine_types.go:53]: The union tags come from an unimplemented KEP from upstream Kube, eventually they should be implemented

File: api/v1beta1/openstackmachine_types.go
Comment (JoelSpeed) [inline api/v1beta1/openstackmachine_types.go:68]: Minimum or maximum values that are applicable here?

File: api/v1beta1/openstackmachine_types.go
Comment (JoelSpeed) [inline api/v1beta1/openstackmachine_types.go:73]: Is the empty string a valid value? Currently it will pass validation without a minLength of 1

Can you impost a maximal string length to this?

File: api/v1beta1/openstackmachine_types.go
Comment (JoelSpeed) [inline api/v1beta1/openstackmachine_types.go:82]: Add an explicit `+kubebuilder:validation:Required` tag please

File: api/v1alpha1/openstackserver_types.go
Comment (JoelSpeed) [inline api/v1alpha1/openstackserver_types.go:108]: Would be useful to describe to end users what they might be able to achieve with this field so that when they `kubectl explain` it tells them what they can do

File: api/v1beta1/openstackmachine_types.go
Comment (JoelSpeed) [inline api/v1beta1/openstackmachine_types.go:55]: Should explain ideally that the valid values are `Bool`, `String` and `Number`. Your statement includes them but it would be better to be direct and explicit
```suggestion
	// Type represents the type of the value.
	// Valid values are Bool, String, and Number.
```

**Comment by EmilienM** [comment]
/lgtm cancel
Valid comments from Joel (thanks for taking the time).

File: api/v1beta1/openstackmachine_types.go
Comment (mdbooth) [reply api/v1beta1/openstackmachine_types.go:68]: I don't think there can be any minima or maxima in this case, as the nature of the interface is we don't know what it will be used for.

I wonder if we should make the size explicit anyway by making this an int64.

**Review by mdbooth — COMMENTED** [review]
Please can you rebase on top of https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/2173 when it lands to fix the bizarre error executing the -test job.

When you do that it's going to highlight a couple of genuine issues. There are some lint errors and some failures in apivalidations.

File: api/v1beta1/openstackmachine_types.go
Comment (JoelSpeed) [reply api/v1beta1/openstackmachine_types.go:68]: You can set the maximum to the maximum of an int64 so that it is at least in the schema, but, ideally it would be lower.

All integer types in Kube APIs should be int32 or int64, other int primitives are not allowed by convention

Comment (okozachenko1203) [reply api/v1beta1/openstackmachine_types.go:68]: I can see that openapi codegen set `int32` as default format for `int` typed values in zz_generated.openapi.go schema.
As long as the schema set it as int32, do you still need to set max or min value using the max and min value of int32 explicitly?

Comment (JoelSpeed) [reply api/v1beta1/openstackmachine_types.go:69]: You should always use the typed int, so, int32 or int64 in this case. And its best practice to have maximums, but in this case, I don't know that it's possible without excluding actually valid OSP API values

**Review by mdbooth — COMMENTED** [review]
Nice work, thanks!

/lgtm

**Pull request merged**
